src/Modules/Tags.elm

```diff
@@ -0,0 +1,352 @@
+    { name : String
+    , value : String
+    }
+
+
+decoder : Decode.Decoder Tag
+decoder =
+    Decode.succeed Tag
+        |> Decode.required "name" Decode.string
+        |> Decode.optional "value" Decode.string ""
+
+
+initialModel : { tags : Dict String String, isInherited : Bool } -> Model
+initialModel { tags, isInherited } =
+    { tags =
+        tags
+            |> Dict.toList
+            |> List.indexedMap
+                (\index value ->
+                    ( index
+                    , { name = Tuple.first value
+                      , value = Tuple.second value
+                      }
+                    )
+                )
+            |> Dict.fromList
+    , newTagName = ""
+    , newTagValue = ""
+    , isInherited = isInherited
+    , newTagError = Nothing
+    , error = Dict.empty
+    }
+
+
+type Msg
+    = NoOp
+    | NameChanged Int { name : String }
+    | ValueChanged Int { value : String }
+    | Removed Int
+    | TagAdded
+    | NewTagNameChanged String
+    | NewTagValueChanged String
+    | TagValidated Int
+    | Submitted
+    | Closed
+
+
+update : Msg -> Model -> { onSubmit : Model -> msg, onClose : msg, onFocus : msg } -> ( Model, Cmd msg )
+update msg model config =
+    case msg of
+        NoOp ->
+            ( model, Cmd.none )
+
+        NameChanged key { name } ->
+            ( { model
+                | tags =
+                    Dict.update
+                        key
+                        (Maybe.map (\a -> { name = name, value = a.value }))
+                        model.tags
+              }
+            , Cmd.none
+            )
+
+        ValueChanged key { value } ->
+            ( { model
+                | tags =
+                    Dict.update
+                        key
+                        (Maybe.map (\a -> { name = a.name, value = value }))
+                        model.tags
+              }
+            , Cmd.none
+            )
+
+        Removed key ->
+            ( { model
+                | tags = Dict.remove key model.tags
+                , error = Dict.remove key model.error
+              }
+            , Cmd.none
+            )
+
+        TagAdded ->
+            let
+                tagAlreadyExists : Bool
+                tagAlreadyExists =
+                    model.tags
+                        |> Dict.toList
+                        |> List.map (Tuple.second >> .name)
+                        |> List.filter ((==) model.newTagName)
+                        |> List.head
+                        |> Maybe.map (\_ -> True)
+                        |> Maybe.withDefault False
+            in
+            ( if tagAlreadyExists then
+                { model | newTagError = Just "exists" }
+
+              else if String.isEmpty model.newTagName then
+                { model | newTagError = Just "empty" }
+
+              else if String.length model.newTagName > 32 then
+                { model | newTagError = Just "too long (32)" }
+
+              else
```

Comment:
It feels like we're implementing the tag validation twice (second time is https://github.com/scrive/elm-challenge/pull/16/files#diff-edc7374f17c064fedb4f48d570e9455210c16d868744151dd8cf1d725c3f011fR180 )

This is a maintenance problem.

Reply:
Sure it is legal :) This task is thought as being as close to a regular day to day job in Scrive as possible and as such part of the discussion about the code in PR are some changes on stuff that we find out can be improved.

Although just do the changes in bulk and ideally once as we need to cover more candidates and it might become little bit messy and we might lost track of what has been changed and when.